k-distribution/src/main/scripts/lib/pyk/kastManip.py

```diff
@@ -508,9 +590,377 @@ def removeSourceMap(k):
+                    newConstraints.append(c)
+                    usedVars.extend(newVars)
+        usedVars = list(set(usedVars))
+    return mlAnd([state] + newConstraints)
+
+
+def removeConstraintClausesFor(varNames, constraint):
+    constraints = flattenLabel('#And', constraint)
+    newConstraints = []
+    for c in constraints:
+        if not any([v in varNames for v in collectFreeVars(c)]):
+            newConstraints.append(c)
+    return mlAnd(newConstraints)
+
+
+def removeConstraintsFor(varNames, constrainedTerm):
+    (state, constraint) = splitConfigAndConstraints(constrainedTerm)
+    constraint = removeConstraintClausesFor(varNames, constraint)
+    return mlAnd([state, constraint])
+
+
+def hasExistentials(pattern):
+    return any([v.startswith('?') for v in collectFreeVars(pattern)])
+
+
+def buildRule(ruleId, initConstrainedTerm, finalConstrainedTerm, claim=False, priority=None, keepVars=None):
+    (initConfig, initConstraint) = splitConfigAndConstraints(initConstrainedTerm)
+    (finalConfig, finalConstraint) = splitConfigAndConstraints(finalConstrainedTerm)
+    initConstraints = flattenLabel('#And', initConstraint)
+    finalConstraints = [c for c in flattenLabel('#And', finalConstraint) if c not in initConstraints]
+    initConstrainedTerm = mlAnd([initConfig] + initConstraints)
+    finalConstrainedTerm = mlAnd([finalConfig] + finalConstraints)
+
+    lhsVars = collectFreeVars(initConstrainedTerm)
+    rhsVars = collectFreeVars(finalConstrainedTerm)
+    varOccurances = countVarOccurances(mlAnd([initConstrainedTerm, finalConstrainedTerm]))
+    vSubst = {}
+    vremapSubst = {}
+    for v in varOccurances:
+        newV = v
+        if varOccurances[v] == 1:
+            newV = '_' + newV
+        if v in rhsVars and v not in lhsVars:
+            newV = '?' + newV
+        vSubst[v] = KVariable(newV)
+        vremapSubst[newV] = KVariable(v)
+
+    initConstrainedTerm = substitute(initConstrainedTerm, vSubst)
+    finalConstrainedTerm = applyExistentialSubstitutions(substitute(finalConstrainedTerm, vSubst))
+    (initConfig, initConstraint) = splitConfigAndConstraints(initConstrainedTerm)
+    (finalConfig, finalConstraint) = splitConfigAndConstraints(finalConstrainedTerm)
+
+    ruleBody = pushDownRewrites(KRewrite(initConfig, finalConfig))
+    ruleRequires = simplifyBool(unsafeMlPredToBool(initConstraint))
+    ruleEnsures = simplifyBool(unsafeMlPredToBool(finalConstraint))
+    ruleAtt = None if claim or priority is None else KAtt(atts={'priority': str(priority)})
+    if not claim:
+        rule = KRule(ruleBody, requires=ruleRequires, ensures=ruleEnsures, att=ruleAtt)
+    else:
+        rule = KClaim(ruleBody, requires=ruleRequires, ensures=ruleEnsures, att=ruleAtt)
+    rule = addAttributes(rule, {'label': ruleId})
+    newKeepVars = None
+    if keepVars is not None:
+        newKeepVars = [vSubst[v]['name'] for v in keepVars]
+    return (minimizeRule(rule, keepVars=newKeepVars), vremapSubst)
```

Comment:
So this will eventually be refactored to: https://github.com/runtimeverification/k/pull/2441#discussion_r808393671

I guess we should get the downstream repos working with this reorganization first, then we can get the improvements to code merged.